### PR TITLE
drivers: sensor: tmp112: Add default device initialization

### PR DESCRIPTION
**Pull request by kurddt**
This allow to have a smaller driver footprint when a single
configuration is required.

Allow to have multiple driver instances

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

### CODE REVIEW

File: drivers/sensor/tmp112/tmp112.c
Comment (nashif) [inline drivers/sensor/tmp112/tmp112.c:197]: The code here seems to be against an older version of Zephyr, the Kconfigs and some of the APIs have changed, please update against latest Zephyr tree.

**Review by nashif — DISMISSED** [review]

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:197]: I didn't saw that is config has been changed in favor to `CONFIG_PM_DEVICE`. I will also change it in my other open PR

File: drivers/sensor/tmp112/Kconfig
Comment (gmarull) [inline drivers/sensor/tmp112/Kconfig:37]: This setting should be part of devicetree, not Kconfig

File: drivers/sensor/tmp112/Kconfig
Comment (gmarull) [inline drivers/sensor/tmp112/Kconfig:52]: Same here

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:24]: This function is redundant, `struct tmp112_config *config = dev->config` can be used instead

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:29]: Same here

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:199]:
```suggestion
			     uint32_t *state, pm_device_cb cb, void *arg)
```

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:244]:
```suggestion
		LOG_ERR("device %s not ready!", config->bus->name);
```

**Review by gmarull — CHANGES_REQUESTED** [review]
Thanks for these changes, added some comments.

File: drivers/sensor/tmp112/Kconfig
Comment (kurddt) [reply drivers/sensor/tmp112/Kconfig:37]: True this would allow to have per device init values

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:24]: I used to have compilation warning when doing so let me check again

File: drivers/sensor/tmp112/Kconfig
Comment (avisconti) [reply drivers/sensor/tmp112/Kconfig:37]: Of course in DT to allow multi-instance

File: drivers/sensor/tmp112/tmp112.c
Comment (avisconti) [reply drivers/sensor/tmp112/tmp112.c:24]: I'm not against the usage of inline functions like this, honestly. The code looks more clear and compiler should optimize anyway. But if you want to follow the suggestion I think the warning issue is maybe because the "const" dropped.

**Review by avisconti — CHANGES_REQUESTED** [review]
I see 3 different purposes in the same commit. Shouldn't it be splitted into at least 3 different commits with all suggestions received?

1. Add multi-instance support
2. Add fixed configuration at compile time
3. Add PM routines

What do you thi nk?

**Comment by kurddt** [comment]
> I see 3 different purposes in the same commit. Shouldn't it be splitted into at least 3 different commits with all suggestions received? 
>     1. Add multi-instance support
>     2. Add fixed configuration at compile time 
>     3. Add PM routines 
> 
> What do you think?

I started doing that but since all three modify the newly introduced file `tmp112.h` it was not so straight forward to doing so. If you think It's a must have I can have a second attempt at it. It will probably be easier when I will switch from KConfig to DT

**Comment by avisconti** [comment]
> > I see 3 different purposes in the same commit. Shouldn't it be splitted into at least 3 different commits with all suggestions received?
> > 1. Add multi-instance support
> > 2. Add fixed configuration at compile time
> > 3. Add PM routines
> > What do you think?
> 
> I started doing that but since all three modify the newly introduced file `tmp112.h` it was not so straight forward to doing so. If you think It's a must have I can have a second attempt at it. It will probably be easier when I will switch from KConfig to DT

Yes please. Let's give a try,

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:24]: I also like those accessor function. they are present in some other device drivers as well

File: drivers/sensor/tmp112/tmp112.c
Comment (avisconti) [inline drivers/sensor/tmp112/tmp112.c:106]: Is this part (and related call) belonging to multi-instance commit? Or is it part of the configuration change?

File: drivers/sensor/tmp112/tmp112.c
Comment (avisconti) [inline drivers/sensor/tmp112/tmp112.c:140]: same this one

File: drivers/sensor/tmp112/Kconfig
Comment (avisconti) [inline drivers/sensor/tmp112/Kconfig:21]: Putting this part here means that once selected all instance will be configurable at runtime. Is this what you want? I guess it is done for simplicity, and I'm not against it.

File: drivers/sensor/tmp112/tmp112.c
Comment (avisconti) [inline drivers/sensor/tmp112/tmp112.c:231]: I was expecting tmp112_pm_control to be part of the driver definition.

**Review by avisconti — CHANGES_REQUESTED** [review]
PLease take care also of putting the commit messages correctly (the multi-instance one is not correct)

File: drivers/sensor/tmp112/Kconfig
Comment (kurddt) [reply drivers/sensor/tmp112/Kconfig:21]: Yes that's true. This is done to safe some resources (flash) when the sensor configuration does not need to be changed after initialization which I think is the most common use-case. That said for backward compatibility it would be better to have it activated by default, like the older version. This has been lost in the rebase

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:106]: It's actually a bit of both, I will try to reflect that in the commit separation

**Review by avisconti — APPROVED** [review]
LGTM

File: drivers/sensor/tmp112/Kconfig
Comment (gmarull) [inline drivers/sensor/tmp112/Kconfig:22]: nit: both are missing `help`

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (gmarull) [inline dts/bindings/sensor/ti,tmp112.yaml:23]: nit: `Celsius`

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (gmarull) [inline dts/bindings/sensor/ti,tmp112.yaml:14]: nit: `mHz`

File: drivers/sensor/tmp112/tmp112.h
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.h:48]: this state is not initialized (should likely be set to active on init)

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:212]: as far as I can see, only `PM_DEVICE_STATE_ACTIVE` is handled, if system requests `PM_DEVICE_STATE_SUSPEND` the driver does nothing but changing its state. How does device save power?

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [reply drivers/sensor/tmp112/tmp112.c:24]: I think most of these helpers come from a period when Zephyr had _complicated names_ for the config/data fields in the device struct. I won't block on that, but I think new or refactored drivers should stop using such dummy functions.

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:212]: On it's own it does not. But if the device can be powered on/off by external electronics then the state can be restored.

Comment (gmarull) [reply drivers/sensor/tmp112/tmp112.c:212]: A driver that implements PM needs to support at the very least active/suspend states. If the device can be powered on/off externally, then the device driver needs to know the actual device state, or, have control of such external electronics to change the device power states.

Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:212]: > then the device driver needs to know the actual device state

Comment (gmarull) [reply drivers/sensor/tmp112/tmp112.c:212]: The problem I see here is that this is not actually implementing any device PM. If system is suspended, the device will get called with `PM_DEVICE_STATE_SET` and `PM_DEVICE_STATE_SUSPEND`. In this case, the device will do nothing other than setting it's state to `PM_DEVICE_STATE_SUSPEND`. As far as I understand, this is equivalent to have no device PM.

File: drivers/sensor/tmp112/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/tmp112/Kconfig:21]: please move this line above the help section

File: drivers/sensor/tmp112/Kconfig
Comment (MaureenHelm) [inline drivers/sensor/tmp112/Kconfig:28]: please move this line above the help section

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/ti,tmp112.yaml:16]: Where does this default come from? Is it the hardware reset value in the sensor?

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (MaureenHelm) [inline dts/bindings/sensor/ti,tmp112.yaml:25]: ditto

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:212]: The driver send the last known configuration when the pm state is set from `PM_DEVICE_STATE_SUSPEND` to `PM_DEVICE_STATE_ACTIVE`. I record the current state in order to not send the configuration twice if the `PM_DEVICE_STATE_SET` is done twice in a row. We could save the configuration upon `PM_DEVICE_STATE_SUSPEND` but it's not required since `tmp112_attr_set` already does it

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (kurddt) [reply dts/bindings/sensor/ti,tmp112.yaml:16]: I thought it was but after checking the hardware default is actually 4Hz. I will change that to align with hardware reset values

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (kurddt) [reply dts/bindings/sensor/ti,tmp112.yaml:25]: Yes this is the hardware default

File: drivers/sensor/tmp112/tmp112.h
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.h:48]: done

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:222]: done

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [reply drivers/sensor/tmp112/tmp112.c:203]: I'm sorry, but I still don't understand how the device enters a _low power state_ if the system is suspended. `tmp112_attr_set` has nothing to do with device PM. Can you please detail the workflow?

Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:203]: So let's say we have a board with several power domain. In one of those power domain, controlled by component `PWR`, we can find among other devices a `TMP112` device.

The `PWR` device driver (or the application) do know which components via (struct devices *) are under the power domain it controls. When the power domain is not in use anymore and need to be brought down, `TMP112` `pm_device_state_set(dev, PM_DEVICE_STATE_SUSPEND)` is called and then the power is turned-off.

When the power domain is required, i.e the application need access to one of the component. The power is turned on by `PWR` and  `pm_device_state_set(dev, PM_DEVICE_STATE_ACTIVE)` is called

File: drivers/sensor/tmp112/tmp112.c
Comment (MaureenHelm) [inline drivers/sensor/tmp112/tmp112.c:77]: Is changing the type related to adding multi-instance support? If not, please make this change in a separate commit.

Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:77]: Moved to a separate commit

**Review by MaureenHelm — COMMENTED** [review]
Need to resolve PM questions raised by @gmarull , but otherwise LGTM.

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [reply drivers/sensor/tmp112/tmp112.c:203]: I'd expect to see something related to this sensor feature in the PM section:

image

What you are describing is really an application specific behavior as far as I understand.

**Comment by kurddt** [comment]
@gmarull For some reason Github only allow me to resolve the conversation but not to add a reply. So I will answer here

You are right the shutdown feature could be used in the suspend state.

I don't think the PM implementation I propose is a application specific behavior. 
Without I don't think it's possible to have fine grain device power management with zephyr using device driver. 

That said I can remove the power management function and start a discussion on Slack about how to support such use case if you prefer.

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:218]: an enum property being used as a boolean sounds weird, why not use a boolean in devicetree?

File: drivers/sensor/tmp112/tmp112.h
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.h:31]: cast should be performed outside of the macro

File: drivers/sensor/tmp112/tmp112.c
Comment (gmarull) [inline drivers/sensor/tmp112/tmp112.c:62]: `const` redundant here

File: drivers/sensor/tmp112/Kconfig
Comment (gmarull) [inline drivers/sensor/tmp112/Kconfig:23]: is it really needed to make these kconfig options, considering these are not per-instance?

Comment (kurddt) [reply drivers/sensor/tmp112/Kconfig:23]: This is in order to reduce the application footprint when those features are not required

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:62]: What do you mean ? The `const` is to indicate that `new_val` will not change after being set

File: drivers/sensor/tmp112/Kconfig
Comment (gmarull) [reply drivers/sensor/tmp112/Kconfig:23]: ok, understood, thanks for the clarification

File: drivers/sensor/tmp112/tmp112.c
Comment (kurddt) [reply drivers/sensor/tmp112/tmp112.c:218]: That way the devicetree and `tmp112_attr_set(SENSOR_ATTR_FULL_SCALE)` are coherent, they accept the same values

**Comment by kurddt** [comment]
The PR will now probably conflicts now that #37750 has been recently merged

**Comment by kurddt** [comment]
@MaureenHelm @gmarull Could you approve the PR if everything looks good to you ?

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (gmarull) [inline dts/bindings/sensor/ti,tmp112.yaml:22]: this property is being treated as a boolean (stored in `extended_mode`, which is `bool extended_mode: 1`). DT binding should reflect that as well.

File: dts/bindings/sensor/ti,tmp112.yaml
Comment (gmarull) [inline dts/bindings/sensor/ti,tmp112.yaml:13]: driver uses `cr` (conversion rate), can names be aligned?

**Review by gmarull — CHANGES_REQUESTED** [review]

**Comment by kurddt** [comment]
@gmarull I addressed your latest requested changes

**Review by MaureenHelm — APPROVED** [review]

**Review by gmarull — APPROVED** [review]

**Comment by kurddt** [comment]
@nashif I would be nice if you can review again this PR in order to be merged

**Pull request merged**
